### PR TITLE
Network solver tool updates

### PR DESCRIPTION
**Pull request by simonseyock**
This PR introduces the following features:
* A context menu entry to clear the active group
* The possibility to move any point in the point digitize mode
* The possibility to delete any point in the point digitize mode

It also contains a refactoring that was necessary to provide this features.

Resolves: #252
Resolves: #253
Resolves: #88

### CODE REVIEW

**Comment by simonseyock** [comment]
@geographika could you check if this provides the desired features? The delete is done by holding the ctrl key while clicking. It works the same with polygons.

The possibility to configure this is still missing. I would suggest one config option `featuresModifiable` that would turn on modification and deleting for both points and polygons.

Also circle are currently not possible to modify and it is not possible to remove a polygon entirely.

@geographika @chrismayer This PR is quite bulky to review. I suggest to either review the commit https://github.com/compassinformatics/cpsi-mapview/pull/317/commits/ef9b32455a2685b5da1aa4789177a3e7d78fdca4 beforehand or I can seperate this PR into 4 different units (clear group, refactoring, moving, deleting).

**Comment by simonseyock** [comment]
To give a better visual indication of the delete it would be nice to have a special delete style in red that would be applied to the feature if hovered while the ctrl key is pressed. Is this desired?

**Comment by simonseyock** [comment]
I think I made a conceptual error in understanding the network solver tool working with points. It calculates the shortest graph that contains all the given points. The order of the given points does not play a role. Is that correct?

**Comment by geographika** [comment]
@simonseyock - the order is important - it solves from one point to the next based on the point `index`. Overlaps are then removed by the service. 
I've emailed @verendi with API doc links.

**Comment by simonseyock** [comment]
@geographika thanks for the info. Based on that I was able to implement everything properly (At least I hope I did XD)

The functionality can already be reviewed, for the code I do some rebasing for easier reviewing

File: app/controller/button/DigitizeButtonController.js [hunk 1]
```diff
@@ -243,21 +283,163 @@ Ext.define('CpsiMapview.controller.button.DigitizeButtonController', {
...
+     */
+    handlePointDelete: function (evt) {
+        var me = this;
+
+        var features = me.map.getFeaturesAtPixel(evt.pixel, {
+            layerFilter: function (layer) {
+                return layer === me.drawLayer;
+            }
+        });
+        if (features.length) {
```
Comment (geographika) [inline app/controller/button/DigitizeButtonController.js:419]: `features` could be null here and throws an error (happened during testing)

File: app/controller/button/DigitizeButtonController.js [hunk 2]
```diff
@@ -243,21 +283,163 @@ Ext.define('CpsiMapview.controller.button.DigitizeButtonController', {
...
+    handleModifyEnd: function (evt) {
+        var me = this;
+        var view = me.getView();
+
+        // find modified feature
+        var drawFeature = me.map.getFeaturesAtPixel(evt.mapBrowserEvent.pixel, {
+            layerFilter: function (layer) {
+                return layer === me.drawLayer;
+            }
+        })[0];
```
Comment (geographika) [inline app/controller/button/DigitizeButtonController.js:366]: `drawFeature` could be null (happened during testing)

**Review by geographika — COMMENTED** [review]
@simonseyock - thanks for this - it provides nice improvement to the UX. 

Some comments:

- the context-menu should always be available with a "Clear All" option. When groups are used the "Clear All" option can remain at the bottom to remove all features from all groups. "Clear All" should be available for all the different types (polygon, circle, line). 

image

The circle tool keeps adding new features. Maybe the code at #313 needs to be (re)applied?
image

I'm not sure I fully understand the deletion process. In the screencast at https://drive.google.com/file/d/1oPuIprkI4k9j_uDd-qqtd0gG7IldOObk/view I held CTRL and clicked the point but the whole path was removed. I was using Chrome on Windows if that makes a difference. 
In terms of approach, would it be possible to add a "Delete Point" option to the context menu when right-clicking on a point? Then the deletion could be controlled with just the mouse.

File: app/controller/button/DigitizeButtonController.js (hunk 2 above)
Comment (simonseyock) [reply app/controller/button/DigitizeButtonController.js:366]: did this happen during deletion of a polygon vertex or any other modification? If it happened during any other modification I have to investigate further.

**Comment by simonseyock** [comment]
I fixed the deletion and the other points.
Only for the context menu I need a clarification: At the moment it is controlled by the `useContextMenu` option. Should this option get removed then?

**Comment by simonseyock** [comment]
I just keep the option for now.

**Comment by simonseyock** [comment]
Should the groups also be available for circles and polygons? If that is the case I would suggest two options `groups` and `clear` instead of the option `useContextMenu`.

Also there is the option `clearDrawnFeature` which does not make really sense in my opinion. In which case would you want to keep your drawn features? I would remove it because it is not really clear what it would do for example in case of modified or deleted points.

**Comment by geographika** [comment]
> 
> 
> I fixed the deletion and the other points.
> Only for the context menu I need a clarification: At the moment it is controlled by the `useContextMenu` option. Should this option get removed then?

Yes this must seem a little confusing. Only the point solver uses groups, which is set by `useContextMenu`. It would make more sense to have a separate `useGroups` option.

**Comment by geographika** [comment]
> Should the groups also be available for circles and polygons? If that is the case I would suggest two options `groups` and `clear` instead of the option `useContextMenu`.

`groups` and `clear` sound a good approach.
Groups are not used for circles and polygons currently. 

> Also there is the option `clearDrawnFeature` which does not really make sense in my opinion. In which case would you want to keep your drawn features? I would remove it because it is not really clear what it would do for example in case of modified or deleted points.

In `CpsiMapview.view.toolbar.MapTools` you'll see this option is set to `false` for polygons and circles so the feature is retained after the lines are returned. 

image

**Comment by simonseyock** [comment]
> In CpsiMapview.view.toolbar.MapTools you'll see this option is set to false for polygons and circles so the feature is retained after the lines are returned.

Ok I thought this was the default. What should the option do for points then? Behave like modifiable was set to false?

**Comment by geographika** [comment]
> 
> 
> > In CpsiMapview.view.toolbar.MapTools you'll see this option is set to false for polygons and circles so the feature is retained after the lines are returned.
> 
> Ok I thought this was the default. What should the option do for points then? Behave like modifiable was set to false?

Not sure I fully understand the question. 
It is just the point tool that needs to hide the "drawn" points as they are also returned by the service.

**Comment by geographika** [comment]
@simonseyock - let me know if the above answer your questions. I tried the latest branch but this doesn't seem to have addressed the comments in https://github.com/compassinformatics/cpsi-mapview/pull/317#pullrequestreview-513569640 - I'm not sure if these have been added or are yet to be pushed.

**Comment by simonseyock** [comment]
@geographika the context menu is not modified yet, because I had these questions. The circles should work, though.

About the options: I am in general way more in favour of providing only options that are really needed because with every new option you generate new edge cases where different options are combined that you did not think of beforehand. That's why asked all these questions :) In general I think sensible defaults are preferable. But for now I will implement with the 3 options.

**Comment by geographika** [comment]
There are 3 variations of the tool. Each tool involves 2 layers:

1. The user "draw" layer, which contains the features created by the user
2. The "results" layer which contains line features sent back from the server. 

The 3 tools are:

- the point solver tool. Any of the points created by the user in the "draw" should not be retained once the "results" layer is populated. Any previous "results" should be cleared, unless they are in another group. This is the only solver which works with groups at the moment. 
- the polygon solver tool. With this tool the last polygon created by the user in the "draw" layer should be retained to show the outline around the "results". Every time a user creates a new polygon both the "draw" layer and "results" layer should be cleared. 
- the line solver currently has no implementation and is not used. 

If only 2 options are required to implement these then all the better.

**Comment by simonseyock** [comment]
@geographika  You are forgetting the circle selection.

In my opinion there is no need for any of theses options - everything could depend on the tool type:
If it is `Point` provide groups, a `clear active group` option and make them modifiable - because why not? The modify interaction should not interfere with the normal usage.
If it is `Circle` or `Polygon` provide a `clear all` option. If it is `Polygon` also make it modifiable (this is currently the behaviour).

**Comment by simonseyock** [comment]
And always do not clear the drawn feature for `Circle` or `Polygon`

**Comment by simonseyock** [comment]
I added a commit that provides the options, though.

**Comment by geographika** [comment]
> 
> 
> @geographika You are forgetting the circle selection.
> 
> In my opinion there is no need for any of theses options - everything could depend on the tool type:
> If it is `Point` provide groups, a `clear active group` option and make them modifiable - because why not? The modify interaction should not interfere with the normal usage.
> If it is `Circle` or `Polygon` provide a `clear all` option. If it is `Polygon` also make it modifiable (this is currently the behaviour).

Yes the options aren't really required - all is driven by the tool type. Sorry I missed the Circle tool - this should behave like the polygon tool. All but the last entered user feature are removed. 
Yes all features should be modifiable.

**Comment by simonseyock** [comment]
@geographika Just to be clear: Should I then remove the options `useContextMenu`, `clearDrawnFeature` and also not add the new options `modifiable`, `groups`, `clear`?

**Comment by geographika** [comment]
> 
> 
> @geographika Just to be clear: Should I then remove the options `useContextMenu`, `clearDrawnFeature` and also not add the new options `modifiable`, `groups`, `clear`?

I think defaults can be used based on the tool type rather than having options - with the exception of `goups` - this would apply to the point solver only.

**Comment by simonseyock** [comment]
@geographika The options are removed, all your remarks should be resolved.

@chrismayer Could you review this code? It maybe makes sense to review the commit https://github.com/compassinformatics/cpsi-mapview/pull/317/commits/d7a65654c80613f125d5e324df3603f126de785c first which should not add any new functionality and the rest afterwards: https://github.com/compassinformatics/cpsi-mapview/pull/317/files/d7a65654c80613f125d5e324df3603f126de785c..bee8297cdbbc42ec6182eaac3a083a25b7815afb

**Review by chrismayer — APPROVED** [review]

**Comment by geographika** [comment]
@simonseyock - works very well. Thanks for all the updates. I'm happy to merge this (I see @chrismayer has approved) so will do so in the next hour.

**Pull request merged**